File: apps/spanwork/src/lib/tauri/project_category.ts
/**
 * 项目分类 CRUD IPC 封装（project_category_* commands）
 *
 * 供 ProjectCategoriesPage 与 CategorySelect 下拉使用。
 */
import type {
  CreateProjectCategoryInput,
  ProjectCategoryDto,
  UpdateProjectCategoryInput,
} from '@spanwork/shared-types';

import { tauriInvoke } from './client';

export function listProjectCategories(): Promise<ProjectCategoryDto[]> {
  return tauriInvoke<ProjectCategoryDto[]>('project_category_list');
}

export function createProjectCategory(
  input: CreateProjectCategoryInput,
): Promise<ProjectCategoryDto> {
  return tauriInvoke<ProjectCategoryDto>('project_category_create', { input });
}

export function updateProjectCategory(
  id: string,
  patch: UpdateProjectCategoryInput,
): Promise<ProjectCategoryDto> {
  return tauriInvoke<ProjectCategoryDto>('project_category_update', {
    params: { id, patch },
  });
}

export function deleteProjectCategory(id: string): Promise<void> {
  return tauriInvoke<void>('project_category_delete', { id });
}

export function reorderProjectCategories(orderedIds: string[]): Promise<void> {
  return tauriInvoke<void>('project_category_reorder', {
    params: { orderedIds },
  });
}

File: apps/spanwork/src-tauri/src/commands/project_category.rs
//! 项目分类 CRUD 与排序 IPC。
//! 委托 db/repos/project_category，删除分类时会解除关联项目的 category_id。

use tauri::State;

use crate::db::repos::project_category as category_repo;
use crate::dto::{
    CreateProjectCategoryInput, ProjectCategoryDto, ProjectCategoryReorderParams,
    ProjectCategoryUpdateParams,
};
use crate::error::AppResult;
use crate::state::AppState;

#[tauri::command]
pub fn project_category_list(state: State<'_, AppState>) -> AppResult<Vec<ProjectCategoryDto>> {
    state.with_db("project_category_list", category_repo::list)
}

#[tauri::command]
pub fn project_category_create(
    state: State<'_, AppState>,
    input: CreateProjectCategoryInput,
) -> AppResult<ProjectCategoryDto> {
    state.with_db("project_category_create", |conn| category_repo::create(conn, &input))
}

#[tauri::command]
pub fn project_category_update(
    state: State<'_, AppState>,
    params: ProjectCategoryUpdateParams,
) -> AppResult<ProjectCategoryDto> {
    state.with_db("project_category_update", |conn| {
        category_repo::update(conn, &params.id, &params.patch)
    })
}

#[tauri::command]
pub fn project_category_delete(state: State<'_, AppState>, id: String) -> AppResult<()> {
    state.with_db("project_category_delete", |conn| category_repo::delete(conn, &id))
}

#[tauri::command]
pub fn project_category_reorder(
    state: State<'_, AppState>,
    params: ProjectCategoryReorderParams,
) -> AppResult<()> {
    state.with_db("project_category_reorder", |conn| category_repo::reorder(conn, &params))
}
